R/weight.R

```diff
@@ -59,3 +59,106 @@ setMethod("weightVariables", "VariableCatalog", function (x) {
         return(c())
     }
 })
+
+#' Generate a weight vector
+#'
+#' This function allows you to generate a weight vector by supplying a set of
+#' categorical variables and the target distribution for each of the variable's categories.
+#' For instance if you wanted to create a weight variable which equally weighted four categories stored
+#' in `ds$var` you would call `makeWeight(ds$var ~ c(25, 25, 25, 25), name = "weight")`.
+#'
+#' @param ...
+#' A series of expressions of the form `variable ~ target_weights`. The variable must
+#' be a categorical crunch variable, and the target weights must be a numeric vector whose
+#' length is equal to the number of categories contained in the variable, and whose sum is equal to 100.
+#'
+#' @param name
+#' The name of the resulting variable
+#' @rdname makeWeight
+#' @return
+#' A crunch Variable Definiton of the weight variable
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' mtcars$cyl <- as.factor(mtcars$cyl)
+#' mtcars$gear <- as.factor(mtcars$gear)
+#' ds <- newDataset(cars, "cars_fact")
+#' ds$weight <- makeWeight(ds$cyl ~ c(30, 30, 40, 0), ds$gear ~ c(20, 20, 60, 0), name = "weight" )
+#' as.vector(ds$weight)
+#' }
+makeWeight <- function(..., name) {
+    expr_list <- list(...)
+    lapply(expr_list, validateWeightExpression)
+    out <- list(
+        name = name,
+        derivation = list(
+            `function` = "rake",
+            args = lapply(expr_list, generateWeightEntry))
+        )
+    class(out) <- "VariableDefinition"
+    out
+}
+
+#' Validate an expression passed to makeWeight
+#'
+#' Utility function to catch formula errors passed to [makeWeight].
+#'
+#' @param expr
+#' An expression
+#' @return
+#' NULL value, function is called for its side effects
+#'
+#' @keywords internal
+#' @examples
+#' \dontrun{
+#' validateWeightExpression(ds$var ~ c(10, 30, 60))
+#' }
+#'
+validateWeightExpression <- function(expr) {
+    if (length(expr) != 3) {
```

Comment:
This should be ensuring that the expression is a formula and should use R's formula-parsing utilities, I think. cf. https://github.com/Crunch-io/rcrunch/blob/master/R/formula.R#L11-L17